File: solutions/scala/bottle-song/src/main/scala/BottleSong.scala
object BottleSong:
  private val numbers = List("no", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten")

  private def suffix(bottles: Int) = if bottles == 1 then "" else "s"

  private def hanging(bottles: Int) =
    s"${numbers(bottles)} green bottle${suffix(bottles)} hanging on the wall"

  private def verse(bottles: Int) =
    Seq(
      hanging(bottles) + ",",
      hanging(bottles) + ",",
      "And if one green bottle should accidentally fall,",
      "There'll be " + hanging(bottles - 1).toLowerCase
    ).mkString("", "\n", ".\n")


  def recite(startBottles: Int, takeDown: Int): String =
    startBottles until startBottles - takeDown by -1 map verse mkString "\n"

File: solutions/python/bottle-song/bottle_song.py
NUMBERS = ("no", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten")


def hanging(bottles):
    suffix = "" if bottles == 1 else "s"
    return f"{NUMBERS[bottles]} green bottle{suffix} hanging on the wall"


def verse(bottles):
    current_hanging = hanging(bottles)
    next_hanging = hanging(bottles - 1).lower()
    return (
        f"{current_hanging},",
        f"{current_hanging},",
        f"And if one green bottle should accidentally fall,",
        f"There'll be {next_hanging}.",
        ""
    )


def recite(start: int, take=1) -> list[str]:
    """Recite verses from a countdown song, starting from a given number for a certain count.

    Args:
        start: The starting verse number to recite from.
        take: The number of verses to recite, taken in reverse order. Defaults to 1.

    Returns:
        A list of strings, where each string contains a line of the recited lyrics.
    """
    return list(line for i in range(start, start - take, -1) for line in verse(i))[:-1]
